File: flink-runtime/src/test/scala/org/apache/flink/runtime/jobmanager/Tasks.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.runtime.jobmanager

import org.apache.flink.runtime.execution.Environment
import org.apache.flink.runtime.io.network.api.reader.RecordReader
import org.apache.flink.runtime.io.network.api.writer.{RecordWriter, RecordWriterBuilder}
import org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable
import org.apache.flink.types.IntValue

object Tasks {

  class Forwarder(environment: Environment)
    extends AbstractInvokable(environment) {

    override def invoke(): Unit = {
      val reader = new RecordReader[IntValue](
        getEnvironment.getInputGate(0),
        classOf[IntValue],
        getEnvironment.getTaskManagerInfo.getTmpDirectories)
      
      val writer = new RecordWriterBuilder().build(
        getEnvironment.getWriter(0)).asInstanceOf[RecordWriter[IntValue]]

      try {
        while (true) {
          val record = reader.next()

          if (record == null) {
            return
          }

          writer.emit(record)
        }

        writer.flushAll()
      } finally {
        writer.clearBuffers()
      }
    }
  }

  class AgnosticReceiver(environment: Environment)
    extends AbstractInvokable(environment) {

    override def invoke(): Unit = {
      val reader= new RecordReader[IntValue](
        getEnvironment.getInputGate(0),
        classOf[IntValue],
        getEnvironment.getTaskManagerInfo.getTmpDirectories)

      while(reader.next() != null){}
    }
  }

  class AgnosticBinaryReceiver(environment: Environment)
    extends AbstractInvokable(environment) {

    override def invoke(): Unit = {
      val reader1 = new RecordReader[IntValue](
        getEnvironment.getInputGate(0),
        classOf[IntValue],
        getEnvironment.getTaskManagerInfo.getTmpDirectories)
      
      val reader2 = new RecordReader[IntValue](
        getEnvironment.getInputGate(1),
        classOf[IntValue],
        getEnvironment.getTaskManagerInfo.getTmpDirectories)

      while(reader1.next() != null){}
      while(reader2.next() != null){}
    }
  }

  class AgnosticTertiaryReceiver(environment: Environment)
    extends AbstractInvokable(environment) {

    override def invoke(): Unit = {
      val env = getEnvironment

      val reader1 = new RecordReader[IntValue](
        env.getInputGate(0),
        classOf[IntValue],
        getEnvironment.getTaskManagerInfo.getTmpDirectories)
      
      val reader2 = new RecordReader[IntValue](
        env.getInputGate(1),
        classOf[IntValue],
        getEnvironment.getTaskManagerInfo.getTmpDirectories)
      
      val reader3 = new RecordReader[IntValue](
        env.getInputGate(2),
        classOf[IntValue],
        getEnvironment.getTaskManagerInfo.getTmpDirectories)

      while(reader1.next() != null){}
      while(reader2.next() != null){}
      while(reader3.next() != null){}
    }
  }

  class ExceptionSender(environment: Environment)
    extends AbstractInvokable(environment) {

    override def invoke(): Unit = {
      throw new Exception("Test exception")
    }
  }

  class ExceptionReceiver(environment: Environment)
    extends AbstractInvokable(environment) {

    override def invoke(): Unit = {
      throw new Exception("Test exception")
    }
  }

  class InstantiationErrorSender(environment: Environment)
    extends AbstractInvokable(environment) {
    throw new RuntimeException("Test exception in constructor")

    override def invoke(): Unit = {
    }
  }

  class BlockingReceiver(environment: Environment)
    extends AbstractInvokable(environment) {
    override def invoke(): Unit = {
      val o = new Object
      o.synchronized(
        o.wait()
      )
    }
  }
}

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/Tasks.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.runtime.jobmanager;

import org.apache.flink.runtime.execution.Environment;
import org.apache.flink.runtime.io.network.api.reader.RecordReader;
import org.apache.flink.runtime.io.network.api.writer.RecordWriter;
import org.apache.flink.runtime.io.network.api.writer.RecordWriterBuilder;
import org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable;
import org.apache.flink.types.IntValue;

import java.io.IOException;

/** Collection of {@link AbstractInvokable}s. */
public class Tasks {

    /** An {@link AbstractInvokable} that forwards all incoming elements. */
    public static class Forwarder extends AbstractInvokable {

        public Forwarder(Environment environment) {
            super(environment);
        }

        @Override
        public void invoke() throws Exception {

            final RecordReader<IntValue> reader =
                    new RecordReader<>(
                            getEnvironment().getInputGate(0),
                            IntValue.class,
                            getEnvironment().getTaskManagerInfo().getTmpDirectories());

            final RecordWriter<IntValue> writer =
                    new RecordWriterBuilder<IntValue>().build(getEnvironment().getWriter(0));

            try {
                while (true) {
                    final IntValue record = reader.next();

                    if (record == null) {
                        return;
                    }

                    writer.emit(record);
                }
            } finally {
                writer.close();
            }
        }
    }

    /** An {@link AbstractInvokable} that consumes 1 input channel. */
    public static class AgnosticReceiver extends AbstractInvokable {

        public AgnosticReceiver(Environment environment) {
            super(environment);
        }

        @Override
        public void invoke() throws Exception {
            consumeInputs(1, this);
        }
    }

    /** An {@link AbstractInvokable} that consumes 2 input channels. */
    public static class AgnosticBinaryReceiver extends AbstractInvokable {

        public AgnosticBinaryReceiver(Environment environment) {
            super(environment);
        }

        @Override
        public void invoke() throws Exception {
            consumeInputs(2, this);
        }
    }

    /** An {@link AbstractInvokable} that consumes 3 input channels. */
    public static class AgnosticTertiaryReceiver extends AbstractInvokable {

        public AgnosticTertiaryReceiver(Environment environment) {
            super(environment);
        }

        @Override
        public void invoke() throws Exception {
            consumeInputs(3, this);
        }
    }

    private static void consumeInputs(int numberOfInputs, AbstractInvokable consumer)
            throws IOException, InterruptedException {
        for (int i = 0; i < numberOfInputs; i++) {
            final RecordReader<IntValue> reader =
                    new RecordReader<>(
                            consumer.getEnvironment().getInputGate(i),
                            IntValue.class,
                            consumer.getEnvironment().getTaskManagerInfo().getTmpDirectories());
            while (reader.next() != null) {}
        }
    }

    /** An {@link AbstractInvokable} that throws an exception when being invoked. */
    public static class ExceptionSender extends AbstractInvokable {

        public ExceptionSender(Environment environment) {
            super(environment);
        }

        @Override
        public void invoke() throws Exception {
            throw new Exception("Test exception");
        }
    }

    /** An {@link AbstractInvokable} that throws an exception when being invoked. */
    public static class ExceptionReceiver extends AbstractInvokable {

        public ExceptionReceiver(Environment environment) {
            super(environment);
        }

        @Override
        public void invoke() throws Exception {
            throw new Exception("Test exception");
        }
    }

    /** An {@link AbstractInvokable} that throws an exception when being instantiated. */
    public static class InstantiationErrorSender extends AbstractInvokable {

        public InstantiationErrorSender(Environment environment) {
            super(environment);
            throw new RuntimeException("Test exception in constructor");
        }

        @Override
        public void invoke() throws Exception {}
    }
}
